File: CalendarView.py
from objc import YES, NO, IBAction, IBOutlet
from Foundation import *
from AppKit import *
import datetime
import calendar
from waker import *
from transformers import IntToNSColor, NSDateFromString

#calendar.setfirstweekday(NSCalendar.currentCalendar().firstWeekday())
def drawRect(r):
    bp = NSBezierPath.bezierPathWithRect_(r)
    color = NSColor.lightGrayColor()
    color.set()
    bp.stroke()

def fillRect(r, color):
    bp = NSBezierPath.bezierPathWithRect_(r)
    color.set()
    bp.fill()

class CalendarView(NSView):
    cache = {}
    
    title = IBOutlet()
    progress = IBOutlet()
    
    def awakeFromNib(self):
        import datetime
        now = datetime.datetime.now()
        self.year = now.year
        self.month = now.month
        self.progress.stopAnimation_(self)
        self.updateTitle()
        self.loading = False
        self.loadingThread = None
        
    def updateTitle(self):
        self.setNeedsDisplay_(YES)
        self.title.setStringValue_('Preview of %s %s' % (month_to_string[self.month], self.year))
    
    @IBAction
    def nextMonth_(self, sender):
        self.month += 1
        if self.month == 13:
            self.year += 1
            self.month = 1
        self.updateTitle()

    @IBAction
    def previousMonth_(self, sender):
        self.month -= 1
        if self.month == 0:
            self.year -= 1
            self.month = 12
        self.updateTitle()
    
    def refresh(self):
        self.cache = {}
        self.setNeedsDisplay_(True)
        
    def endLoading_(self, sender):
        self.progress.stopAnimation_(self)
        self.setNeedsDisplay_(True)

    def loadDataThread_(self, params):
        pool = NSAutoreleasePool.alloc().init()
        self.loadingThread = NSThread.currentThread()
        self.loading = True
        self.progress.startAnimation_(self)
        year, month, month_calendar = params
        for y, week in enumerate(month_calendar):
            for x, day in enumerate(week):
                if day != 0:
                    # make some weak attempt at trying to detect that some other thread has cleared the cache and started another thread
                    if self.loadingThread != NSThread.currentThread():
                        #print 'another loading thread detected, bailing'
                        return
                    key = (year, month, day)
                    self.cache[key] = get_rule(NSApplication.sharedApplication().delegate(), year, month, day)
        self.performSelectorOnMainThread_withObject_waitUntilDone_(self.endLoading_, self, False)
        self.loading = False
    
    def drawRect_(self, in_rect):
        today = datetime.datetime.now()
        year = self.year
        month = self.month
        month_calendar = calendar.monthcalendar(year, month)
        if (year, month, 1) not in self.cache:
            NSThread.detachNewThreadSelector_toTarget_withObject_(self.loadDataThread_, self, (year, month, month_calendar))
            return
        
        weekday_text_height = 15
        in_rect.size.height -= weekday_text_height

        cell_size = NSSize(in_rect.size.width/len(month_calendar[0]), in_rect.size.height/len(month_calendar))
        
        # week titles
        centeredParagraphStyle = NSMutableParagraphStyle.alloc().init()
        centeredParagraphStyle.setAlignment_(NSCenterTextAlignment)
        for x, day in enumerate(weekday_to_string):
            rect = NSRect(
                NSPoint(cell_size.width*x, in_rect.origin.x+in_rect.size.height), 
                NSSize(cell_size.width, weekday_text_height))
            NSString.stringWithString_(weekday_to_string[day].title()).drawInRect_withAttributes_(rect, {NSForegroundColorAttributeName:NSColor.grayColor(), NSParagraphStyleAttributeName:centeredParagraphStyle}) # centered!
            
        in_rect.size.height -= 5
        
        # day data
        for y, week in enumerate(month_calendar):
            for x, day in enumerate(week):
                if day != 0:
                    rect = NSRect(
                        NSPoint(cell_size.width*x, in_rect.origin.x-cell_size.height*(y+1)+in_rect.size.height), 
                        NSSize(cell_size.width, cell_size.height))
                    rect = NSInsetRect(rect, 2, 2)
                    fillRect(rect, NSColor.colorWithDeviceRed_green_blue_alpha_(0.9, 0.9, 1, 1) if (year, month, day) == (today.year, today.month, today.day) else NSColor.whiteColor())

                    key = (year, month, day)
                    try:
                        item = self.cache[key]
                    except KeyError:
                        continue
                    if item:
                        fillRect(NSRect(rect.origin, NSSize(rect.size.width, 20)), IntToNSColor(item.color()))
                    rect.origin.x += 3
                    rect.size.width -= 3
                    NSString.stringWithString_(str(day)).drawInRect_withAttributes_(rect, {NSForegroundColorAttributeName:NSColor.grayColor()})
                    
                    rect.origin.x += 2
                    rect.size.height -= 14
                    
                    if item is not None:
                        if item.name() is not None:
                            NSString.stringWithString_(item.name()).drawInRect_withAttributes_(rect, {NSParagraphStyleAttributeName:centeredParagraphStyle})
                        if item.time() is not None:
                            rect.size.height -= 38
                            rect.size.width -= 4
                            rect.origin.y += 21
                            font = NSFont.fontWithName_size_('Geneva', min(rect.size.height-4, rect.size.width/3))
                            attributes = {NSFontAttributeName:font, NSParagraphStyleAttributeName:centeredParagraphStyle}
                            timeString = NSString.stringWithString_(item.time())
                            size = timeString.sizeWithAttributes_(attributes)
                            rect.origin.y -= rect.size.height/2-size.height/2
                            NSString.stringWithString_(timeString).drawInRect_withAttributes_(rect, attributes)

File: CalendarView.m
#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>
#import "CalendarView.h"
#import "NSDate+NSDate_Humane.h"
#import "transformers.h"
#import "waker.h"
#import "Waker_AppDelegate.h"

static NSDictionary* _weekday_to_string = nil;
static NSDictionary* _month_to_string = nil;

static NSDictionary* weekday_to_string(void);
static NSDictionary* month_to_string(void);

static void fillRect(NSRect r, NSColor* color) {
    NSBezierPath* bp = [NSBezierPath bezierPathWithRect:r];
    [color set];
    [bp fill];
}

static NSString* date_to_key(int year, int month, int date) {
    return [NSString stringWithFormat:@"%d %d %d", year, month, date];
}


static NSArray* monthcalendar(int year, int month) {
    NSDateComponents* comps = [[NSDateComponents alloc] init];
    [comps setDay:1];
    [comps setMonth:month];
    [comps setYear:year];
    
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDate* monthDate = [calendar dateFromComponents:comps];
    
    NSRange r = [calendar rangeOfUnit:NSCalendarUnitDay inUnit:NSCalendarUnitMonth forDate:monthDate];
    int number_of_days_in_month = r.length;
    int weekday = [monthDate weekday];
    NSMutableArray* result = [@[] mutableCopy];
    NSMutableArray* row = nil;
    for (int day = 1; day != number_of_days_in_month+1; day++) {
        if (row == nil) {
            row = [@[] mutableCopy];
            [result addObject:row];
        }
        if (day == 1) {
            // Pad 0s from last month
            for (int x = 0; x != weekday; x++) {
                [row addObject:[NSNumber numberWithInt:0]];
            }
        }
        
        [comps setDay:day];
        NSDate* today = [calendar dateFromComponents:comps];
        [row addObject:[NSNumber numberWithInt:day]];
        if ([today weekday] == 6) {
            row = nil;
        }
    }
    
    return result;
}


NSDictionary* weekday_to_string(void) {
    if (_weekday_to_string == nil) {
        _weekday_to_string = @{
            @0: @"monday",
            @1: @"tuesday",
            @2: @"wednesday",
            @3: @"thursday",
            @4: @"friday",
            @5: @"saturday",
            @6: @"sunday",
        };
    }
    return _weekday_to_string;
}

NSDictionary* month_to_string(void) {
    if (_month_to_string == nil) {
        _month_to_string = @{
            @1: @"January",
            @2: @"February",
            @3: @"March",
            @4: @"April",
            @5: @"May",
            @6: @"June",
            @7: @"July",
            @8: @"August",
            @9: @"September",
            @10: @"October",
            @11: @"November",
            @12: @"December",
        };
    }
    return _month_to_string;
}

@implementation CalendarView : NSView

- (void)awakeFromNib {
    NSDate* now = [NSDate new];
    
    self->_year = now.year;
    self->_month = now.month;
    [self->progress stopAnimation:self];
    [self updateTitle];
    self->loading = false;
    self->loadingThread = nil;
    self->cache = [@{} mutableCopy];
}

- (void)updateTitle {
    [self setNeedsDisplay:YES];
    NSDictionary* d = month_to_string();
    [self->title setStringValue:[NSString stringWithFormat:@"Preview of %@ %d", [d objectForKey:[NSNumber numberWithInt:self->_month]], self->_year]];
}

- (IBAction)nextMonth:(id __unused)sender {
    self->_month += 1;
    if (self->_month == 13) {
        self->_year += 1;
        self->_month = 1;
    }
    [self updateTitle];
}

- (IBAction)previousMonth:(id __unused)sender {
    self->_month -= 1;
    if (self->_month == 0) {
        self->_year -= 1;
        self->_month = 12;
    }
    [self updateTitle];
}

- (void)refresh {
    [self->cache removeAllObjects];
    [self setNeedsDisplay:true];
}

- (IBAction)endLoading:(id __unused)sender {
    [self->progress stopAnimation:self];
    [self setNeedsDisplay:true];
}

//NSRange daysInMonth(int year, int month) {
//    NSDateComponents* comps = [[NSDateComponents alloc] init];
//    [comps setDay:1];
//    [comps setMonth:month];
//    [comps setYear:year];
//    
//    NSCalendar *calendar = [NSCalendar currentCalendar];
//    NSDate* monthDate = [calendar dateFromComponents:comps];
//    return [calendar rangeOfUnit:NSDayCalendarUnit
//                          inUnit:NSMonthCalendarUnit
//                         forDate:monthDate];
//}

- (void)loadDataThread:(NSArray*)params {
    @autoreleasepool {
        self->loadingThread = [NSThread currentThread];
        self->loading = true;
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self->progress startAnimation:self];
        });
        
        int year = [[params objectAtIndex:0] intValue];
        int month = [[params objectAtIndex:1] intValue];
        
        NSArray* month_calendar = monthcalendar(year, month);
        for (int y = 0; y != month_calendar.count; y++) {
            NSArray* week = [month_calendar objectAtIndex:y];
            for (int x = 0; x != week.count; x++) {
                int day = [[week objectAtIndex:x] intValue];
                if (day != 0) {
                    // make some weak attempt at trying to detect that some other thread has cleared the cache and started another thread
                    if (self->loadingThread != [NSThread currentThread]) {
                        //print 'another loading thread detected, bailing'
                        return;
                    }
                    dispatch_sync(dispatch_get_main_queue(), ^{
                        NSString* key = date_to_key(year, month, day);
                        Waker_AppDelegate* app_delegate = (Waker_AppDelegate*)[[NSApplication sharedApplication] delegate];
                        NSManagedObject* rule = get_rule(app_delegate, year, month, day);
                        if (rule == nil) {
                            [self->cache removeObjectForKey:key];
                        }
                        else {
                            [self->cache setObject:rule forKey:key];
                        }
                    });
                }
            }
        }
        [self performSelectorOnMainThread:@selector(endLoading:) withObject:self waitUntilDone:false];
        self->loading = false;
    }
}

- (void)drawRect:(NSRect)in_rect {
    NSDate* now = [NSDate new];

    int year = self->_year;
    int month = self->_month;

    if ([self->cache objectForKey:date_to_key(year, month, 1)] == nil) {
        [NSThread detachNewThreadSelector:@selector(loadDataThread:) toTarget:self withObject:@[[NSNumber numberWithInt:year], [NSNumber numberWithInt:month]]];
        return;
    }
    int weekday_text_height = 15;
    in_rect.size.height -= weekday_text_height;
    NSArray* month_calendar = monthcalendar(year, month);
    NSSize cell_size = NSMakeSize(in_rect.size.width / 7, in_rect.size.height / month_calendar.count);
    // week titles
    NSMutableParagraphStyle* centeredParagraphStyle = [[NSMutableParagraphStyle alloc] init];
    [centeredParagraphStyle setAlignment:NSTextAlignmentCenter];
    for (NSNumber* x in @[@0, @1, @2, @3, @4, @5, @6]) {
        NSString* day = [weekday_to_string() objectForKey:x];
        NSRect rect = NSMakeRect(
                                 cell_size.width * [x intValue],
                                 in_rect.origin.x + in_rect.size.height,
                                 cell_size.width,
                                 weekday_text_height);
        [day drawInRect:rect withAttributes:@{NSForegroundColorAttributeName: [NSColor grayColor], NSParagraphStyleAttributeName: centeredParagraphStyle}];
    }
    in_rect.size.height -= 5;
    // day data
    NSColor *todayColor = isDarkMode()? [NSColor colorWithWhite:1.0 alpha:0.2] : [NSColor colorWithDeviceRed:0.9 green:0.9 blue:1 alpha:1];
    NSColor *otherDayColor = isDarkMode() ? [NSColor colorWithWhite:1.0 alpha:0.05] : [NSColor colorWithWhite:0.0 alpha:0.05];
    for (int y = 0; y != month_calendar.count; y++) {
        NSArray* week = [month_calendar objectAtIndex:y];
        for (int x = 0; x != week.count; x++) {
            int day = [[week objectAtIndex:x] intValue];
            if (day != 0) {
                NSRect rect = NSMakeRect(
                                         cell_size.width * x,
                                         in_rect.origin.x - cell_size.height * (y + 1) + in_rect.size.height,
                                         cell_size.width,
                                         cell_size.height);
                rect = NSInsetRect(rect, 2, 2);
                BOOL is_today = year == now.year && month == now.month && day == now.day;
                fillRect(rect, is_today? todayColor : otherDayColor);
                NSString* key = date_to_key(year, month, day);
                NSManagedObject* item = [self->cache objectForKey:key];
                if (item == nil) {
                    continue;
                }
                fillRect(NSMakeRect(rect.origin.x, rect.origin.y, rect.size.width, 20), IntToNSColor([[item color] intValue]));

                rect.origin.x += 3;
                rect.size.width -= 3;
                [[NSString stringWithFormat:@"%d", day] drawInRect:rect withAttributes:@{NSForegroundColorAttributeName: [NSColor textColor]}];
                rect.origin.x += 2;
                rect.size.height -= 14;
                if (item != nil) {
                    if (([item name] != nil)) {
                        [[NSString stringWithString:[item name]] drawInRect:rect withAttributes:@{NSParagraphStyleAttributeName: centeredParagraphStyle, NSForegroundColorAttributeName: [NSColor textColor]}];
                    }
                    if ([item time] != nil) {
                        rect.size.height -= 38;
                        rect.size.width -= 4;
                        rect.origin.y += 21;
                        NSFont* font = [NSFont fontWithName:@"Geneva" size:MIN(rect.size.height - 4, rect.size.width / 3)];
                        NSDictionary* attributes = @{NSFontAttributeName: font, NSParagraphStyleAttributeName: centeredParagraphStyle, NSForegroundColorAttributeName: [NSColor textColor]};
                        NSString* timeString = [NSString stringWithString:[item time]];
                        NSSize size = [timeString sizeWithAttributes:attributes];
                        rect.origin.y -= rect.size.height / 2 - size.height / 2;
                        [[NSString stringWithString:timeString] drawInRect:rect withAttributes:attributes];
                    }
                }
            }
        }
    }
}

@end
